### PR TITLE
New README

### PR DESCRIPTION
**Pull request by skade**
Documenting our whole teaching offer.

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
 
 The material is created with people with zero Rust experience but with a programming background in mind.
 
-## Overview
+Ferrous Systems offers a large Rust curriculum for both beginner and
+advanced Rust developer at all levels. We are the oldest and most often
+chosen Rust training provider.
+
+Ferrous Systems specialises in custom, topic focused workshops for
+enterprises starting moving into Rust. Supplemental to our courses, we
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
enterprises starting or expanding their use of Rust. Supplemental to our courses, we
```

File: README.adoc [hunk 2]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+offer ongoing help and feedback.
```
Comment (jamesmunns) [inline README.adoc]: Link to Rust-Experts?

File: README.adoc [hunk 3]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+
+Overview
+--------
 
 The materials are presented as a set of small, self-contained lessons on a specific topic.
+Note that lessons might be taken down when deemed outdated and recreated for new audiences.
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Note that lessons might be revised, extended, or removed, when necessary to keep material up to date or relevant to new audiences.
```

File: README.adoc [hunk 4]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
 Lessons can be arranged into courses exploring various aspects of Rust.
 
-## Building
+Pricing
+~~~~~~~
+
+Ferrous Systems usually charges per trainer day. That allows for easy
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Ferrous Systems charges on a per-trainer-per-day basis. That allows for easy
```

File: README.adoc [hunk 5]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+calculation on one-off trainings and a fair and easy discussion about
```
Comment (jamesmunns) [inline README.adoc]: This paragraph seems to exist to justify our pricing scheme, which I don't think is necessary.

I would replace with something like:

> For standard trainings, Ferrous Systems charges on a per-trainer-per-day basis. Quotations for custom trainings are available on request.

File: README.adoc [hunk 6]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+the structure of the course. Other agreements are possible. For example,
+that allows clients to quickly change the setup of the group, e.g. by
+adding a person, without changing the contractual setup.
+
+We offer training on any Rust subject and will create material that we
+deem resuable at no cost for the client.
+
+We also offer training directly on clients products, against a fee to
```
Comment (jamesmunns) [inline README.adoc]: I suggest rewording to:

> We can also tailor training material to be directly relevant for customer's products or domain areas. Please contact us for a quotation for customized training material.

File: README.adoc [hunk 7]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+produce the material.
+
+Group setup
```
Comment (jamesmunns) [inline README.adoc]: I'm not sure if the first and last paragraphs of this section are helpful for our customers to understand our offerings. Perhaps remove paragraph one, and replace paragraph three with:

> When selecting groups for training, we recommend that participants have a similar experience level in Rust, and similar goals for using the language when possible. We are experienced with training developers from a wide range of backgrounds and experience levels, and can design a training suitable for your team.

File: README.adoc [hunk 8]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+to efficiently allow for 2 trainers.
+
+We have experiences with both groups of mixed and uniform backgrounds
+and mixed skill levels. We generally recommend groups of mixed
+backgrounds and similar experience levels.
+
+Training material
+~~~~~~~~~~~~~~~~~
+
+Ferrous Systems offers their training material under MIT/Apache-2.0
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Ferrous Systems offers our training material under MIT/Apache-2.0
```

File: README.adoc [hunk 9]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+license for source code and Creative Commons (CC-BY-SA) for textual
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
license for source code (examples and exercises) and Creative Commons (CC-BY-SA) for textual
```

File: README.adoc [hunk 10]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+content.
+
+Course material specifically constructed for you will be proprietary to
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Custom course material developed on request, including any provided examples or exercises, will be delivered directly as part of the training.
```

File: README.adoc [hunk 11]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+you.
+
+Trainer staff
+~~~~~~~~~~~~~
+
+The Ferrous Systems trainer staff consists of multiple Rust project
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
The Ferrous Systems training staff consists of multiple Rust project
```

File: README.adoc [hunk 12]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+makes sure our trainers are always up to speed on new developments and
+can give current and authoritive answers.
+
+Among those are:
+
+* Florian Gilcher: Training Rust since 2015. Over 30 trainings, most for
+large enterprises. Rust core team member and community team manager.
+Broad knowledge about all aspects of Rust, with special knowledge on
+network and async/concurrent programming.
+* James Munns: Training Rust since 2018. Focus on general Rust and
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
* James Munns: Training Rust since 2017. Focus on general Rust and
```

File: README.adoc [hunk 13]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+small/embedded systems. Member of the Embedded Working Group Core Team.
+Great knowledge of integrating Rust with C codebases.
+* Aleksey Kladov: Training Rust since 2018 in academic and enterprise
+setting. Member of the Rust tools team. Author of the rust-analyzer IDE.
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
settings. Member of the Rust tools team. Author of the rust-analyzer IDE.
```

File: README.adoc [hunk 14]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+especially focused on debugging.
+* Jonas Schievek: Training Rust since 2019. Focus on general Rust and
+multiple architectures. Contributor to the Rust compiler. Great
+knowledge of compilation details and performance concerns.
+* Tanks Transfeld: Training Rust since 2018. Focus on general Rust and
+Rust for beginners. Part of several Rust learning projects. All Rust
+subjects, from async to bare metal. Great knowledge around beginners
+concerns and how to get started.
+
+Ferrous Systems can also collaborate with local trainers in Europe, US,
```
Comment (jamesmunns) [inline README.adoc]: We can probably also throw Canada on the list, or generalize to "North America".

File: README.adoc [hunk 15]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+``Three days'' overview
++++++++++++++++++++++++
+
+A course that focuses more on an ecosystem overview over a detailed core
+subject. It may cover up to 4 topics.
+
+Attendees may not become productive in all areas, but get a good feeling
+for them.
+
+This course is perfect for engineering management, that will most likely
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
This course is perfect for engineering management that will most likely
```

File: README.adoc [hunk 16]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+not end up using Rust daily, but needs broad knowledge to assess Rust as
+a technology. It is also frequently given in academic settings.
+
+5 days
+^^^^^^
+
+The longer versions covers similar subjects as above, but in more
+detail. This allows e.g. the construction of a larger piece of software
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
detail. This allows for exercises such as the construction of a larger piece of software
```

File: README.adoc [hunk 17]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+in the training group collaboratively.
+
+These courses are recommended for people that are intended to be
+multipliers in your organisation.
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
multipliers in your organisation, teaching or leading other teams using Rust.
```

File: README.adoc [hunk 18]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+will be reduced and some self-guided exercises reduced to ``code along''
+sessions.
+
+2,4 and 8 hours topic tasters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Taster sessions that favor experiences over solid knowledge and get
+people excited.
+
+Perfect for conferences and internal all-hands to expand peoples
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Perfect for conferences and internal all-hands to expand people's
```

File: README.adoc [hunk 19]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+horizons.
+
+Advanced courses
+~~~~~~~~~~~~~~~~
+
+Advanced courses deepen a Rust subject and are indended for groups that
+already visited a course before. They quickly refreshen core knowledge
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
have already attended a course previously. They quickly refresh core knowledge
```

File: README.adoc [hunk 20]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+and then add a topical deep dive on topic. Advanced courses also differ
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
and then add a specific deep dive on topic. Advanced courses also differ
```

File: README.adoc [hunk 21]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+from base courses in that they include a number facilitated discussions
+for attendees to share their experience.
+
+Advanced courses are tailored like base courses to the need of the
+group. For example, an basic course for for asynchronous programming
+explains how concurrent execution in Rust works, the advanced one may
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
explains how concurrent execution in Rust works, while the advanced one may
```

File: README.adoc [hunk 22]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+* Overview of common libraries
+* ``Refactoring towards Speed'': a core technique to safely derive a
+fast program from a working program
+
+Exercises will be tailored to the chosen special subjects.
+
+Advanced Generic Programming in Rust
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using libraries with generics in Rust is taught in the core course. When
```
Comment (jamesmunns) [inline README.adoc]: I suggest replacing this paragraph with:

> Although the use of generics is taught in the core course, this section instead focuses on how, and when, to use them effectively. For developers of widely used libraries (internal or public), this material is particularly important.

File: README.adoc [hunk 23]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+* Introduction into advanced programming with generics in Rust
+* Writing generic APIs
+* Impact on compile time, size and runtime speed
+* Useful generics patterns
+* Patterns to avoid
+
+Rust in C/C++ Environments
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rust is often deployed in existing products, especially withing or among
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
Rust is often deployed in existing products, especially within or among
```

File: README.adoc [hunk 24]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+position and use in the language.
+
+In this module, attendees will learn:
+
+* The role of `unsafe`
+* The scope of `unsafe`
+* Do’s and Don’ts of unsafe Rust
+* Introduction into support APIs, like non-null pointers
+* Potential undefined behaviour arising from the use of `unsafe`
+* Checking unsafe Rust vor safety
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
* Checking unsafe Rust for safety
```

File: README.adoc [hunk 25]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+
+Rust Testing
+~~~~~~~~~~~~
+
+This module teaches advanced Rust testing techniques.
+
+In this module, attendees will learn: * Fuzzing of Rust applications *
```
Comment (jamesmunns) [inline README.adoc]: Does this need to be broken into multiple lines?

File: README.adoc [hunk 26]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+demands. It explains techniques to test for performance and optimise for
+speed.
+
+In this module, attendees will learn:
+
+* Tools to analyse speed and memory consumption
+* Programming techniques for faster programs
+* ``Refactoring towards Speed'': futher deepening for optimising working
+code bases
+* Optimising programs for better resource usage
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
* Optimising programs for specific resource usage needs
```

File: README.adoc [hunk 27]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+
+Cross Compilation
+~~~~~~~~~~~~~~~~~
+
+This module is meant for developers targeting many different
+architectures.
+
+In this module, attendees will learn:
+
+* Rusts cross-compile toolchain
```
Comment (jamesmunns) [inline README.adoc]:
```suggestion
* Rust's cross-compile toolchain
```

**Comment by jamesmunns** [comment]
Also, there are a number of `\`\`quotes like this''`. This might need to be fixed.

File: README.adoc [hunk 28]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
 
 The material is created with people with zero Rust experience but with a programming background in mind.
 
-## Overview
+Ferrous Systems offers a large Rust curriculum for both beginner and
+advanced Rust developer at all levels. We are the oldest and most often
+chosen Rust training provider.
```
[the remark this answers is not in the document]
Comment (skade) [reply README.adoc]: Removed it.

File: README.adoc [hunk 29]
```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
...
+
+Ferrous Systems specialises in custom, topic focused workshops for
```
[the remark this answers is not in the document]
Comment (skade) [reply README.adoc]: What's your preference, I generally spell british.

File: README.adoc (hunk 2 above)
Comment (skade) [reply README.adoc]: If it's folded into a course offering, I would generally prefer to not mingle brands.

File: README.adoc (hunk 6 above)
Comment (skade) [reply README.adoc]: Applied manually.

File: README.adoc (hunk 5 above)
Comment (skade) [reply README.adoc]: Applied manually.

File: README.adoc (hunk 7 above)
Comment (skade) [reply README.adoc]: Applied manually.

File: README.adoc (hunk 22 above)
Comment (skade) [reply README.adoc]: Applied manually.

File: README.adoc (hunk 29 above)
[the remark this answers is not in the document]
Comment (jamesmunns) [reply README.adoc]: I generally prefer american english, out of habit :)

But in general, I prefer consistency to one or the other.

File: README.adoc (hunk 25 above)
Comment (skade) [reply README.adoc]: Yep, that's an error. Fixed.

File: README.adoc (hunk 14 above)
Comment (skade) [reply README.adoc]: Added

**Comment by skade** [comment]
Fixed the not-so-smart-quotes.

**Comment by skade** [comment]
@jamesmunns applied all of your suggestions!

**Comment by skade** [comment]
(And pushed the changes ;))

**Comment by skade** [comment]
Hm, seems like GH does not render the ToC correctly, even though I'm doing it by their docs... maybe @matklad has an idea there?

File: README.adoc [hunk 30]
```diff
@@ -1,62 +1,418 @@
...
+* Florian Gilcher: Training Rust since 2015. Over 30 trainings, most for
+large enterprises. Rust core team member and community team manager.
+Broad knowledge about all aspects of Rust, with special knowledge on
+network and async/concurrent programming.
+* James Munns: Training Rust since 2017. Focus on general Rust and
+small/embedded systems. Member of the Embedded Working Group Core Team.
+Great knowledge of integrating Rust with C codebases.
+* Aleksey Kladov: Training Rust since 2018 in academic and enterprise
+settings. Member of the Rust tools team. Author of the rust-analyzer IDE.
+Very broad and exact knowledge of the language and its tooling,
```
Comment (matklad) [inline README.adoc]:
```suggestion
Contributor to Cargo. Very broad and exact knowledge of the language and its tooling.
```

File: README.adoc [hunk 31]
```diff
@@ -1,62 +1,418 @@
...
+especially focused on debugging.
```
Comment (matklad) [inline README.adoc]:
```suggestion

```

File: README.adoc (hunk 29 above)
[the remark this answers is not in the document]
Comment (skade) [reply README.adoc:17]: I'm going to change it for US.
